File: 0004-largest_palindrome_product/main.py
#!/usr/bin/env python

def is_palindrome(number: int) -> bool:
    first = number
    second = 0

    while first > 0:
        digit = int(first % 10)
        second = second * 10 + digit
        first = int(first / 10)

    return number == second


def main():
    digits = 3
    limit = 10 ** digits
    largest = 0

    for i in reversed(range(1, limit)):
        for j in reversed(range(1, i)):
            product = i * j
            if product > largest and is_palindrome(product):
                largest = product

    print(largest)


if __name__ == "__main__":
    main()

File: 0004-largest_palindrome_product/main.rs
fn is_palindrome(number: i32) -> bool {
    let mut first = number;
    let mut second = 0;
    
    while first > 0 {
       let digit = (first % 10) as i32; 
       second = second * 10 + digit;
       first = (first / 10) as i32;
    }
    
    number == second
}

fn main() {
    let digits = 3;
    let limit = 10_i32.pow(digits);
    let mut largest = 0;
    
    for i in (1..limit).rev() {
        for j in (1..i).rev() {
            let product = i * j;
            
            if product > largest && is_palindrome(product) {
                largest = product;
            }
        }
    }
    
    println!("{}", largest);
}
